### PR TITLE
add SaveExternalXRefSpec in incremental link phase.

### PR DESCRIPTION
**Pull request by vwxyzh**

### CODE REVIEW

File: src/Microsoft.DocAsCode.Build.Engine/LinkPhaseHandler.cs [hunk 1]
```diff
@@ -26,18 +26,29 @@ public LinkPhaseHandler(DocumentBuildContext context, TemplateProcessor template
...
+                        Logger.LogVerbose($"Processor {hostService.Processor.Name}: Postbuilding...");
+                        using (new LoggerPhaseScope("Postbuild", LogLevel.Verbose))
+                        {
+                            Postbuild(hostService);
+                        }
                     }
-                }
-            }
+                },
+                maxParallelism);
```
Comment (ansyral) [inline src/Microsoft.DocAsCode.Build.Engine/LinkPhaseHandler.cs]: risky, how about move it to another pr?

File: src/Microsoft.DocAsCode.Build.Engine/LinkPhaseHandlerWithIncremental.cs [hunk 2]
```diff
@@ -169,6 +174,17 @@ private void RegisterUnloadedXRefSpec(IEnumerable<HostService> hostServices)
             }
         }
 
+        private void LoadExternalXRefSpec()
+        {
+            if (CurrentBuildVersionInfo.ExternalXRefSpecFile != null)
```
Comment (ansyral) [inline src/Microsoft.DocAsCode.Build.Engine/LinkPhaseHandlerWithIncremental.cs]: >CurrentBuildVersionInfo [](start = 16, length = 23)

last?

File: src/Microsoft.DocAsCode.Build.Engine/LinkPhaseHandlerWithIncremental.cs [hunk 3]
```diff
@@ -169,6 +174,17 @@ private void RegisterUnloadedXRefSpec(IEnumerable<HostService> hostServices)
             }
         }
 
+        private void LoadExternalXRefSpec()
```
Comment (ansyral) [inline src/Microsoft.DocAsCode.Build.Engine/LinkPhaseHandlerWithIncremental.cs]: >LoadExternalXRefSpec [](start = 21, length = 20)

not called?

File: src/Microsoft.DocAsCode.Build.Engine/LinkPhaseHandlerWithIncremental.cs [hunk 4]
```diff
@@ -273,6 +289,15 @@ private void UpdateXrefMap(IEnumerable<HostService> hostServices)
             }
         }
 
+        private void SaveExternalXRefSpec()
```
Comment (ansyral) [inline src/Microsoft.DocAsCode.Build.Engine/LinkPhaseHandlerWithIncremental.cs]: >SaveExternalXRefSpec [](start = 21, length = 20)

add a unittest?

**Comment by ansyral** [comment]
add a unittest that external xref spec not restored?

File: src/Microsoft.DocAsCode.Build.Engine/Incrementals/BuildInfo.cs [hunk 5]
```diff
@@ -7,6 +7,8 @@ namespace Microsoft.DocAsCode.Build.Engine.Incrementals
     using System.Collections.Generic;
     using System.IO;
 
+    using Newtonsoft.Json;
```
Comment (hellosnow) [inline src/Microsoft.DocAsCode.Build.Engine/Incrementals/BuildInfo.cs]: >Newtonsoft.Json; [](start = 10, length = 16)

seems not used?

File: src/Microsoft.DocAsCode.Build.Engine/LinkPhaseHandlerWithIncremental.cs [hunk 6]
```diff
@@ -319,6 +329,26 @@ where ofi.LinkToPath.StartsWith(IncrementalContext.LastBaseDir)
             return result;
         }
 
+        private void LoadExternalXRefSpec()
+        {
+            if (LastBuildVersionInfo?.ExternalXRefSpecFile != null)
+            {
+                using (var reader = File.OpenText(Path.Combine(LastBuildVersionInfo.BaseDir, LastBuildVersionInfo.ExternalXRefSpecFile)))
+                {
+                    Context.LoadExternalXRefSpec(reader);
```
Comment (ansyral) [inline src/Microsoft.DocAsCode.Build.Engine/LinkPhaseHandlerWithIncremental.cs]: question, is it correct to restore all externalxrefspec from last?

Comment (vwxyzh) [inline src/Microsoft.DocAsCode.Build.Engine/LinkPhaseHandlerWithIncremental.cs]: Not resolved, it only contains external xref specs created in save phases.

File: src/Microsoft.DocAsCode.Build.Engine/DocumentBuildContext.cs [hunk 7]
```diff
@@ -118,6 +118,24 @@ public void ReportExternalXRefSpec(XRefSpec spec)
...
+        internal void SaveExternalXRefSpec(TextWriter writer)
+        {
+            JsonUtility.Serialize(writer, ExternalXRefSpec);
+        }
+
+        internal void LoadExternalXRefSpec(TextReader reader)
+        {
+            if (ExternalXRefSpec.Count > 0)
+            {
+                throw new InvalidOperationException("Cannot load after report external xref spec.");
```
Comment (hellosnow) [inline src/Microsoft.DocAsCode.Build.Engine/DocumentBuildContext.cs]: >report [](start = 71, length = 6)

minor: reporting

**Comment by ansyral** [comment]
:shipit:

**Comment by superyyrrzz** [comment]
:shipit:
